DESKTOP: Fixed (#2467) edit menu items and shortcuts not disabled in preview mode

Runo-saduwa opened the pull request:
**DESCRIPTION:**
This patch is a fix for the issue #2467 . Previously before this solution, the Edit menu items were not disabled on PREVIEW, this caused users to be able to use short cuts such as CTRL+B, CTRL+` on PREVIEW mode which is an unwanted behavior.

**SOLUTION:**
A function which simulates a switch was created for enabling and disabling these edit menu items, and this function is called conditionally depending on if the panes are in viewer or editor mode. it also takes account of when a notebook/note is empty, it maintains the behavior of disabling these edit menu items/shortcuts when a note has no content.

**HOW TO TEST**
Use the following test cases to confirm this solution
- **Create an empty notebook and click the edit button**: 
    EXPECTED RESULT:  'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be disabled
- **Add a note to a notebook and click the layout toggler to PREVIEW**: 
    EXPECTED RESULT:  'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be disabled
- **Add a note to a notebook and click the layout toggler to EDITOR**: 
    EXPECTED RESULT:  'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be enabled
- **Add a note to a notebook and click the layout toggler to EDITOR AND PREVIEW MODE**: 
    EXPECTED RESULT:  'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be enabled

tessus commented:
@Runo-saduwa please stop requesting reviews from all people in the list. There's no need to request a review at all. We'll look at it when we get to it.

tessus commented on ElectronClient/app/app.js:
```diff
@@ -1148,6 +1149,22 @@ class Application extends BaseApplication {
+			if (!menuItem) continue;
+			menuItem.enabled = true;
+		}
+	}
```
No reason to duplicate this code.

Use one function with a boolean argument. e.g. `enableShortcuts(enable = true)`.

Please use `Shortcuts` not `ShortCuts`. Additionally I suggest to use a global Array for `['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']` since it is also used in `updateMenuItemStates`. It's never a good idea to duplicate code.

Runo-saduwa replied:
@tessus Thanks for the review, I’ll make these changes as soon as possible but please can you explain what you mean by global Array?, how do I do that ?

tessus replied:
Well, maybe global array is the wrong word for OO code. Define it so that it is accessible from both functions (`enableShortcuts` and `updateMenuItemStates`). We don't want the same definition in two different methods.
Then you can do something like: `for (const itemId of menuitemsArray) { ... }`

Runo-saduwa replied:
Okay I get it now, that’s really cool.

laurent22 replied:
This block of code is copied verbatim from updateMenuItemStates. We can't have business logic duplicated three times like that, that's a nightmare to maintain. As a matter of fact, doing that you've broken the logic to disable edit buttons in HTML notes.

Runo-saduwa replied:
I have refactored and I’ve corrected this

tessus replied:
Nope, If you have a HTML note open (in edit mode, or split view) the edit menu items are still ensbled.
A HTML note is created via the web clipper.

laurent22 replied:
Yes it's still the same issue. There are two problems:

1. You've copied and pasted the logic from [here](https://github.com/laurent22/joplin/pull/2529/files#diff-00b0e649e4e574cdf55197c857ccadefL1160) so now there's two places where we need to describe how items are enabled or disabled. Before you copied that logic twice, now it's better because you've copied it only once, but it's still one too many. You need to modify or refactor the existing logic to implement what you want.

2. As tessus said, the logic for HTML notes is broken because you've duplicated the business logic so now the two functions override each other behaviour. Have a look at the code you've copied, please try to understand what it does, and see how you can re-use it.

Runo-saduwa replied:
@tessus thanks, sorry I didn’t test for HTML notes so I didn’t see this, how do I create and open an HTML note

Runo-saduwa replied:
From all Laurent has said, it means I’ll have to find a way to solve this issue using just the updateMenuItemsState function

Runo-saduwa replied:
I’ll work on it and let you know the updates

tessus replied:
To create a HTML note, you have to use the clipper and pick `Clip complete page (HTML) (Beta)`. The resulting note will be in the format HTML. You can verify that by clicking on the note properties toolbar icon. Next to the field `Markdown` it will say `HTML`.

Runo-saduwa replied:
Alright thanks

tessus replied:
> From all Laurent has said, it means I’ll have to find a way to solve this issue using just the updateMenuItemsState function

Yes, you will have to change the `updateMenuItemsState` function to fit your needs. Maybe add a second parameter with a default value. So, the current use of the function doesn't change. e.g. `enable = null`. If enable is null, do what it does now. And if it is true/false do something else. This is just an idea. Maybe there's a better way. You will have to figure this out.

Runo-saduwa replied:
Thanks for the guidance, this is so helpful!

tessus replied:
You are welcome. Btw, you should rebase. The directory `app` does not exist anymore.
Just in case you don't know git to well, make a backup of your branch and/or the `app.js` file.
Laurent changed the build process. The BUILD.md was updated accordingly.

Runo-saduwa replied:
Thanks, I’ll take action.

Runo-saduwa commented:
@tessus I’m sorry, thanks for correcting me

On Thu, Feb 20, 2020 at 12:05 AM Helmut K. C. Tessarek <
notifications@github.com> wrote:

> @Runo-saduwa <https://github.com/Runo-saduwa> please stop requesting
> reviews from all people in the list. There's no need to request a review at
> all. We'll look at it when we get to it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/laurent22/joplin/pull/2529?email_source=notifications&email_token=AIR5ZV2XP4GABIOVHFF5U7DRDW3MBA5CNFSM4KYC7BW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMKBWYA#issuecomment-588520288>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIR5ZV4VHERQ7OZTHLAYE7LRDW3MBANCNFSM4KYC7BWQ>
> .
>

tessus commented:
> I’m sorry, thanks for correcting me

No worries. It's just that we already get a great number of emails.

tessus commented:
You've requested a review from all people again. Please don't do that. I think I told you not to do this in another PR.

Anyway, have you fixed the broken html stuff Laurent mentioned? You haven't repled to his comment.

Runo-saduwa commented:
My apologies, believe me I didn’t mean to break any rule again, I’m still trying to get a hang of the GitHub web interface. Please pardon me this last time. 


On Laurent’s request, I’ve gotten rid of the duplicated business logic and I’ve also made the changes you asked for

PackElend commented:
> My apologies, believe me I didn’t mean to break any rule again, I’m still trying to get a hang of the GitHub web interface. Please pardon me this last time.

pull me in for anything like this, I try to have challenges like this this reflected in the documentation 

> I think I told you not to do this in another PR.

where, so I know what to write :)

Runo-saduwa commented:
For some unknown reason, web clipper is unable to find the port for connecting to my Joplin electron desktop app. i’ve followed all instructions for troubleshooting here https://joplinapp.org/clipper/ but none seems to fit my needs. so i couldn't confirm if this new code works for HTML notes.

But i've removed all duplicate code and modified the `updateMenuStateItems` function

Runo-saduwa commented:
Please something is terribly wrong with this PR, i will have to close it and open a new one

tessus commented:
@Runo-saduwa yep. using git is more than doing a `pull` and `push`. Yo7u are not the only one who has problems with it. Please learn how to use git properly. In the worst case scenario you'll lose all your work, if you don't.

PackElend commented:
> @Runo-saduwa yep. using git is more than doing a `pull` and `push`.
@Runo-saduwa have look here https://zulip.readthedocs.io/en/latest/git/index.html. Zulip has quite a good guide

tessus commented:
any update here on the new PR?

Runo-saduwa commented:
Its ready

Runo-saduwa commented:
Opening a new one now
